File: med/cpp/Question2461.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    long long maximumSubarraySum(vector<int> &nums, int k) {
        long long ans = 0, sum = 0;
        unordered_map<int, int> map;
        for (int i = 0; i < nums.size(); i++) {
            map[nums[i]]++;
            sum += nums[i];
            if (i < k - 1) {
                continue;
            }
            if (map.size() == k) {
                ans = max(ans, sum);
            }
            map[nums[i - k + 1]]--;
            if (map[nums[i - k + 1]] == 0) {
                map.erase(nums[i - k + 1]);
            }
            sum -= nums[i - k + 1];
        }
        return ans;
    }
};

File: python/med/Question2461.py
from collections import defaultdict
from typing import List


class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        ans = 0
        temp_sum = 0
        window = defaultdict(int)
        for i in range(n):
            temp_sum += nums[i]
            window[nums[i]] += 1
            if i - k + 1 < 0:
                continue
            else:
                if len(window) == k:
                    ans = max(ans, temp_sum)
                temp_sum -= nums[i - k + 1]
                window[nums[i - k + 1]] -= 1
                if window[nums[i - k + 1]] == 0:
                    del window[nums[i - k + 1]]
        return ans


print(Solution().maximumSubarraySum([1, 5, 4, 2, 9, 9, 9], 3))
